src/main/scala/firrtl/transforms/DeadCodeElimination.scala

```diff
@@ -0,0 +1,251 @@
+            depGraph.addEdge(node, LogicNode(ext.name, pname))
+          case Port(_, pname, Output, _) => depGraph.addEdge(LogicNode(ext.name, pname), node)
+          case Port(_, pname, Input, _) => depGraph.addEdge(node, LogicNode(ext.name, pname))
+        }
+    }
+    // Connect circuitSink to top-level outputs (and Analog inputs)
+    val topModule = c.modules.find(_.name == c.main).get
+    val topOutputs = topModule.ports.foreach {
+      case Port(_, name, Output, _) => depGraph.addEdge(circuitSink, LogicNode(c.main, name))
+      case Port(_, name, _, AnalogType(_)) => depGraph.addEdge(circuitSink, LogicNode(c.main, name))
+      case _ =>
+    }
+
+    depGraph
+  }
+
+  private def deleteDeadCode(instMap: collection.Map[String, String],
+                             deadNodes: Set[LogicNode],
+                             moduleMap: collection.Map[String, DefModule])
+                            (mod: DefModule): Option[DefModule] = {
+    // Gets all dependencies and constructs LogicNodes from them
+    // TODO this is a duplicate from setupDepGraph, remove by improving how we lookup
+    def getDeps(expr: Expression): Seq[LogicNode] =
+      extractRefs(expr).map { e =>
+        if (kind(e) == InstanceKind) {
+          val (inst, tail) = Utils.splitRef(e)
+          LogicNode(instMap(inst.name), tail)
+        } else {
+          LogicNode(mod.name, e)
+        }
+      }
+
+    var emptyBody = true // TODO is there a better way to do this?
+
+    // TODO Delete unused writers from DefMemory???
+    def onStmt(stmt: Statement): Statement = {
+      val stmtx = stmt match {
+        case inst: WDefInstance =>
+          moduleMap.get(inst.module) match {
+            case Some(instMod) => inst.copy(tpe = Utils.module_type(instMod))
+            case None => EmptyStmt
+          }
+        case decl: IsDeclaration =>
+          val node = LogicNode(mod.name, decl.name)
+          if (deadNodes.contains(node)) EmptyStmt else decl
+        case con: Connect =>
+          val node = getDeps(con.loc) match { case Seq(elt) => elt }
+          if (deadNodes.contains(node)) EmptyStmt else con
+        case Attach(info, exprs) => // If any exprs are dead then all are
+          val dead = exprs.flatMap(getDeps(_)).forall(deadNodes.contains(_))
+          if (dead) EmptyStmt else Attach(info, exprs)
+        case block: Block => block map onStmt
+        case other => other
+      }
+      stmtx match { // Check if module empty
+        case EmptyStmt | _: Block =>
+        case other => emptyBody = false
+      }
+      stmtx
+    }
+
+    val portsx = mod.ports.filterNot(p => deadNodes.contains(LogicNode(mod.name, p.name)))
+
+    mod match {
+      case Module(info, name, _, body) =>
+        val bodyx = onStmt(body)
+        if (emptyBody && portsx.isEmpty) None else Some(Module(info, name, portsx, bodyx))
+      case ext: ExtModule =>
+        if (portsx.isEmpty) None else Some(ext.copy(ports = portsx))
```

Comment:
It is now checked. Actually ran into this bug when I was making extmodules removed only when specified to be removed.